### ISSUE TITLE
Iconify doesn't work

### ISSUE THREAD

**Issue by xinyazhang** [Alexey-Yakovenko/deadbeef#1606]
Version: 0.7.2-2
Environment: Fedora 24, Wayland
Procedure to reproduce: enable 'Close minimizes to tray' and click the close button, deadbeef player simple disappears.

Possible reason:
I took a look at the minimization (or iconify) code. It calls gtk_widget_hide when minimizing to tray is enable. However the Gtk APIs designed for this function are
```
void
gtk_window_iconify (GtkWindow *window);
```
and
```
void
gtk_window_deiconify (GtkWindow *window);
```

**Comment by xinyazhang** [comment]
By the way, the tray icon never shows up ("Hide system tray icon" is enabled). However, thunderbird shows the icon correctly.

I'm looking into this issue.

**Comment by xinyazhang** [comment]
Apparently this is a Wayland-specific issue. Under X11 works without problems.

**Comment by Alexey-Yakovenko** [comment]
> I took a look at the minimization (or iconify) code. It calls gtk_widget_hide when minimizing to tray is enable. However the Gtk APIs designed for this function are 

It's because when an app "minimizes" to sys tray -- it doesn't really minimize or iconify.

I don't know much about wayland and how its system tray implementation works..

Basically, your problem is not what you wrote.. it's that there's no tray icon at all.

The old GTK status icon API is deprecated, so relatively recently deadbeef got a new plugin https://github.com/Alexey-Yakovenko/deadbeef/tree/master/plugins/statusnotifier

This is only available on master branch, so this may explain why it doesn't work if you're using 0.7.2

**Comment by xinyazhang** [comment]
Thanks for you reply.

After some research I found the real problem is Wayland has removed system tray completely (Gnome leaves the tray exclusively for X11 apps).

It seems the proposed solutions by DEs (at least Gnome) is to support MPRIS2 [Quickly Access Media Controls](https://help.gnome.org/misc/release-notes/3.20/), and at least Rhythmbox is working in this way. Thus shall we change this issue as "Mainline MPRIS2 plugin"?

Thank you.

**Comment by Alexey-Yakovenko** [comment]
They are 2 different things.

There's still notification area in KDE5, GNOME3, etc, and it works in deadbeef with the new plugin that I mentioned.

Then there's the Ubuntu sound menu, and similar in other desktops, which are not supported by Deadbeef natively, but there are multiple third party plugins and scripts, which have been available for quite a while.

I'm not convinced to support MPRIS stuff as part of my tree yet, but feel free to post a feature request about it.

**Comment by xinyazhang** [comment]
I totally agree. They are really two different things, but the problem is, from my testing the KDE plugin doesn't work under Gnome 3, while the third-party mpris2 plugin has no problem in presenting status in Gnome.

![statusnotify without mpris2](https://cloud.githubusercontent.com/assets/11429462/16569968/cb60a778-4202-11e6-8461-7ed8fc20f499.png)

![mpris2 with statusnotify disabled](https://cloud.githubusercontent.com/assets/11429462/16569991/0e87d13e-4203-11e6-9521-87764bfe43fc.png)

After taking a look at the statusnotify code, it seems this plugin employs KDE specific D-Bus interfaces ("org.kde.StatusNotifierItem"), and there isn't corresponding interface in Gnome 3 yet.

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang I just tried in GNOME3 which ships in Ubuntu 16.04, and the standard GtkStatusIcon works just fine. It appears in the bottom left corner, like this:

![image](https://cloud.githubusercontent.com/assets/1895698/16581269/dac56770-42a8-11e6-8bc4-3204bf0a35db.png)

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang 

I can't find any information, whether this notification area is still present in wayland version of GNOME shell, or what kind of protocol it's using.

**Comment by Alexey-Yakovenko** [comment]
@ugilio Hey mate, hope it's all right that I'm summoning you.

I tried to figure out why the statusnotifier plugin doesn't work in Ubuntu (Unity)...

Can't find any reason for that. I tried adding more properties to XML, like this:

```
        "    <property type='s' name='IconThemePath' access='read'/>"
        "    <property type='o' name='Menu' access='read'/>"
        "    <property type='b' name='ItemIsMenu' access='read'/>"
```

and handling them like this (hack-ish):

```
    else if (g_strcmp0(property_name, "IconThemePath") == 0)
        ret = g_variant_new_string("/usr/share/icons/hicolor");
    else if (g_strcmp0(property_name, "Menu") == 0)
        ret = g_variant_new_object_path ("/org/deadbeef/StatusNotifier/Menu");
    else if (g_strcmp0(property_name, "ItemIsMenu") == 0)
        ret = g_variant_new_boolean(FALSE);
```

Still couldn't get it to work.

Can you give me a pointer how to debug this stuff? Are there any tools to monitor what's going and where it fails?

**Comment by Alexey-Yakovenko** [comment]
@ugilio I found this in dbus-monitor output:

```
method call time=1467720327.895196 sender=:1.33 -> destination=org.freedesktop.DBus serial=178 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=StartServiceByName
   string "org.kde.StatusNotifierItem-23958-1"
   uint32 0
error time=1467720327.895237 sender=org.freedesktop.DBus -> destination=:1.33 error_name=org.freedesktop.DBus.Error.ServiceUnknown reply_serial=178
   string "The name org.kde.StatusNotifierItem-23958-1 was not provided by any .service files"
```

does this mean that Unity doesn't support this protocol at all, or is it something wrong on Deadbeef's side?

**Comment by Alexey-Yakovenko** [comment]
more info...

```
$ tail -f ~/.cache/upstart/indicator-application.log

(process:1781): indicator-application-service-WARNING **: Notification Item on object /StatusNotifierItem of org.kde.StatusNotifierItem-29924-1 doesn't have enough properties.
```

apparently, I'm still missing some required properties

**Comment by Alexey-Yakovenko** [comment]
Ahh.. I'm wrong, this message disappears when I add the IconThemePath, Menu and ItemIsMenu.

So I'm on the right way, but still not quite there..
Probably need to actually have some sort of Menu object initialized, but didn't figure out how to do this yet.

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang btw, could you please try if this plugin works with wayland?

It uses the same protocol as the built-in status notifier, but seems to be more complete, and works in Ubuntu/Unity at least.

I suppose it should work in GNOME too.

**Comment by Alexey-Yakovenko** [comment]
I tracked it down to the Menu property.. it needs to be a valid GObject path, with an associated DBUS service, which serves the menu, and this could be implemented with the library called "dbusmenu-glib", which is quite a bad dependency to bring.

**Comment by Alexey-Yakovenko** [comment]
I suppose we can repurpose this issue for tracking Unity progress with statusnotifier plugin...

**Comment by xinyazhang** [comment]
@Alexey-Yakovenko 
I believe Gnome has removed system tray icon support for all Wayland Applications, and there is also [a bug report for Lantern](https://github.com/getlantern/lantern/issues/3618) for this problem. This design has been [proposed long time ago](https://wiki.gnome.org/Design/Whiteboards/StatusIcons), but it looks they tried to keep the tray icon working for all X11 applications, for the sake of compatibility, and drop the support for Wayland applications from the beginning.

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang so there's no system tray at all in GNOME shell on wayland?

Can anyone of people who has access to a system running wayland + GNOME confirm?

**Comment by xinyazhang** [comment]
No, I tried vovochka404's plugin, but the status icon still doesn't present.

**Comment by Alexey-Yakovenko** [comment]
I feel like stealing this thread for an unrelated issue (Unity appindicator) was wrong.
Will make a new issue, and close this.

**Comment by xinyazhang** [comment]
@Alexey-Yakovenko 
The actual problem is, we need some method similar with "minimizing to tray" to let deadbeef run in background. If neither mpris2 nor status icon failed to present (that's the case of the git master without 3rd party plugins), after "minimizing" the main UI will simply disappear and there is no way to get it back besides running it again.

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang 

> The actual problem is, we need some method similar with "minimizing to tray" to let deadbeef run in background.

No, we don't. There's no tray in GNOME. Deadbeef can still run "in background" by hiding the main window, but this will remain disabled by default.

Then, the GNOME guidelines explicitly say that apps should NOT do that:

> Applications can run without a window under GNOME 3. However, they are only recommended to do so when:
> The application does not play sound or music.

**Comment by sljunkie** [comment]
I think GNOME + Wayland ignores the system tray completely; the GNOME dev's call it deprecated.

On the other hand, the mpris2 plugin works wonders with deadbeef + GNOME 3 since media player that support it are fully integrated on the notification area (see screenshot). Perhaps deadbeef could include this plugin by default.

![screenshot from 2017-02-02 05-03-57](https://cloud.githubusercontent.com/assets/7532854/22540048/0bca9036-e905-11e6-9a2a-7dd89419e42e.png)

**Comment by AlexanderZhirov** [comment]
Unfortunately, the icon does not appear in the system tray of the Budgie desktop environment.

![2024-02-04_00-28](https://github.com/DeaDBeeF-Player/deadbeef/assets/37277056/7a55aaa0-a482-4d31-b310-a07638112ce0)

**Comment by mozo78** [comment]
I just tried it on KDE with Wayland and there's no systray icon. On X11 it's working fine.

**Comment by mhandze** [comment]
> I just tried it on KDE with Wayland and there's no systray icon. On X11 it's working fine.

Same here. Ubuntu 25.10 (with Wayland as default platform)
